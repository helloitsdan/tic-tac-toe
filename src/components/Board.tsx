import React, { FunctionComponent } from 'react'
import { BoardPosition } from '../types'

interface BoardProps {
  onPositionSelect: (position: BoardPosition) => void
}

const Board: FunctionComponent<BoardProps> = ({
  onPositionSelect,
}) => (
  <div>
    <button onClick={() => onPositionSelect(BoardPosition.TOP_LEFT)}>Top left</button>
    <button onClick={() => onPositionSelect(BoardPosition.TOP_CENTRE)}>Top centre</button>
    <button onClick={() => onPositionSelect(BoardPosition.TOP_RIGHT)}>Top right</button>
    <br />
    <button onClick={() => onPositionSelect(BoardPosition.MIDDLE_LEFT)}>Middle left</button>
    <button onClick={() => onPositionSelect(BoardPosition.MIDDLE_CENTRE)}>Middle centre</button>
    <button onClick={() => onPositionSelect(BoardPosition.MIDDLE_RIGHT)}>Middle right</button>
    <br />
    <button onClick={() => onPositionSelect(BoardPosition.BOTTOM_LEFT)}>Bottom left</button>
    <button onClick={() => onPositionSelect(BoardPosition.BOTTOM_CENTRE)}>Bottom centre</button>
    <button onClick={() => onPositionSelect(BoardPosition.BOTTOM_RIGHT)}>Bottom right</button>
  </div>
)

export default Board
