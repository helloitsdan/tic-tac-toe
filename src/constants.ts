import { BoardPosition, MatchName } from './types'

export const BOARD: BoardPosition[] = [
  BoardPosition.TOP_LEFT,
  BoardPosition.TOP_CENTRE,
  BoardPosition.TOP_RIGHT,
  BoardPosition.MIDDLE_LEFT,
  BoardPosition.MIDDLE_CENTRE,
  BoardPosition.MIDDLE_RIGHT,
  BoardPosition.BOTTOM_LEFT,
  BoardPosition.BOTTOM_CENTRE,
  BoardPosition.BOTTOM_RIGHT,
]

export const POSSIBLE_MATCHES: Record<keyof typeof MatchName, BoardPosition[]> = {
  TOP_ROW: [BoardPosition.TOP_LEFT, BoardPosition.TOP_CENTRE, BoardPosition.TOP_RIGHT],
  MIDDLE_ROW: [BoardPosition.MIDDLE_LEFT, BoardPosition.MIDDLE_CENTRE, BoardPosition.MIDDLE_RIGHT],
  BOTTOM_ROW: [BoardPosition.BOTTOM_LEFT, BoardPosition.BOTTOM_CENTRE, BoardPosition.BOTTOM_RIGHT],
  LEFT_COLUMN: [BoardPosition.TOP_LEFT, BoardPosition.MIDDLE_LEFT, BoardPosition.BOTTOM_LEFT],
  CENTRE_COLUMN: [BoardPosition.TOP_CENTRE, BoardPosition.MIDDLE_CENTRE, BoardPosition.BOTTOM_CENTRE],
  RIGHT_COLUMN: [BoardPosition.TOP_RIGHT, BoardPosition.MIDDLE_RIGHT, BoardPosition.BOTTOM_RIGHT],
  DIAGONAL_LEFT_TO_RIGHT: [BoardPosition.TOP_LEFT, BoardPosition.MIDDLE_CENTRE, BoardPosition.BOTTOM_RIGHT],
  DIAGONAL_RIGHT_TO_LEFT: [BoardPosition.TOP_RIGHT, BoardPosition.MIDDLE_CENTRE, BoardPosition.BOTTOM_LEFT],
}