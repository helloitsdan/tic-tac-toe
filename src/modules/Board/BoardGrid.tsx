import React, { FunctionComponent } from 'react'
import { ReactComponent as BoardIcon } from '../../assets/board.svg'

const BoardGrid: FunctionComponent = () => (
  <BoardIcon className="c-game-board__grid" />
)

export default BoardGrid
