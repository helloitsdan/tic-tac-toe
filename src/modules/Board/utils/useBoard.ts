import { useCallback, useMemo } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { getPlayers, getStatus, getMoves } from '../../../store/selectors'
import { BoardPosition, GameStatus, GameMovesType } from '../../../types'
import isValidMove from './isValidMove'
import { placePiece } from '../../../store/actions'
import getBoardPositionsFromMoves from './getBoardPositionsFromMoves'

const allowMove = (gameStatus: GameStatus, moves: GameMovesType, position: BoardPosition) => {
  if (gameStatus !== GameStatus.PLAYING) {
    return false
  }

  return isValidMove(position, moves)
}

const useBoard = () => {
  const dispatch = useDispatch()
  const { active: activePlayer } = useSelector(getPlayers)
  const gameStatus = useSelector(getStatus)

  const moves = useSelector(getMoves)
  const boardPositions = useMemo(() => getBoardPositionsFromMoves(moves), [moves])

  const onPositionSelect = useCallback(
    (position: BoardPosition) => {
      if (!allowMove(gameStatus, moves, position)) {
        return
      }

      dispatch(placePiece(activePlayer, position))
    },
    [activePlayer, gameStatus, moves, dispatch]
  )

  return [boardPositions, onPositionSelect] as const
}

export { allowMove }
export default useBoard
