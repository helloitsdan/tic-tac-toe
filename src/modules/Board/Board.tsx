import React, { FunctionComponent } from 'react'

import { BOARD } from '../../constants'
import BoardSquare from '../../components/BoardSquare'
import BoardGrid from './BoardGrid'
import useBoard from './utils/useBoard'

const Board: FunctionComponent = () => {
  const [boardPositions, onPositionSelect] = useBoard()

  return (
    <div className="c-game-board">
      <BoardGrid />

      {BOARD.map(position => (
        <div key={position} className="c-game-board__square" data-testid="Board/Square">
          <BoardSquare onSelect={() => onPositionSelect(position)} player={boardPositions[position]} />
        </div>
      ))}
    </div>
  )
}

export default Board
