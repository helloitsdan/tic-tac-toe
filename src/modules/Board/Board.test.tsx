import React from 'react'
import { render } from '@testing-library/react'

import Board from './Board'
import { BOARD } from '../../constants'
import useBoard from './utils/useBoard'

jest.mock('./utils/useBoard')

describe('Board', () => {
  beforeEach(() => {
    const useBoardMock = useBoard as jest.Mock
    useBoardMock.mockReturnValue([{}, jest.fn()])
  })

  it('renders', () => {
    const { asFragment } = render(<Board />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders a square for every board position', () => {
    const { queryAllByTestId } = render(<Board />)
    const boardPositions = queryAllByTestId('Board/Square')
    expect(boardPositions.length).toEqual(BOARD.length)
  })
})
