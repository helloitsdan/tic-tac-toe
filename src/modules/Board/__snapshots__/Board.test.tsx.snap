// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders 1`] = `
<DocumentFragment>
  <div
    class="c-game-board"
  >
    <svg
      class="c-game-board__grid"
    >
      board.svg
    </svg>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
    <div
      class="c-game-board__square"
      data-testid="Board/Square"
    >
      <button
        class="c-board-button"
      >
        <span>
           
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
